ws-server/src/future.rs

```diff
@@ -0,0 +1,132 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any
+// person obtaining a copy of this software and associated
+// documentation files (the "Software"), to deal in the
+// Software without restriction, including without
+// limitation the rights to use, copy, modify, merge,
+// publish, distribute, sublicense, and/or sell copies of
+// the Software, and to permit persons to whom the Software
+// is furnished to do so, subject to the following
+// conditions:
+//
+// The above copyright notice and this permission notice
+// shall be included in all copies or substantial portions
+// of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+// ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT
+// SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
+// IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! Utilities for handling async code.
+
+use futures_util::future::FutureExt;
+use std::future::Future;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+/// This is a flexible collection of futures that need to be driven to completion
+/// alongside some other future, such as connection handlers that need to be
+/// handled along with a listener for new connections.
+///
+/// In order to `.await` on these futures and drive them to completion, call
+/// `select_with` providing some other future, the result of which you need.
+pub(crate) struct FutureDriver<F> {
+	futures: Vec<F>,
+}
+
+impl<F> Default for FutureDriver<F> {
+	fn default() -> Self {
+		FutureDriver { futures: Vec::new() }
+	}
+}
+
+impl<F> FutureDriver<F> {
+	/// Get the count of remaining futures on this driver
+	pub(crate) fn count(&self) -> usize {
+		self.futures.len()
+	}
+
+	/// Add a new future to this driver
+	pub(crate) fn add(&mut self, future: F) {
+		self.futures.push(future);
+	}
+}
+
+impl<F> FutureDriver<F>
+where
+	F: Future + Unpin,
+{
+	pub(crate) async fn select_with<S: Future>(&mut self, selector: S) -> S::Output {
+		tokio::pin!(selector);
```

Comment:
tokio is brought in a dependency just for `pin`?

is there any difference between `futures::pin` and `futures::pin`?

Reply:
I had a gander, and both `tokio::pin` and `futures::pin_mut` have identical impls for this usage; the tokio version just supports an alternate usage as well taht isn't useful here.